src/collectors/ceph/ceph.py

```diff
@@ -69,78 +169,357 @@ def get_default_config(self):
+
+        # Calculate deltas since last time we queried admin socket. The
+        # derivitive function records from the last invocation the
+        # total_sum/total_count, and simply returns the difference.
+        delta_sum = self.derivative(delta_sum_name, total_sum,
+                                    time_delta=False)
+        delta_count = self.derivative(delta_count_name, total_count,
+                                      time_delta=False)
+
+        # average in the last collection interval
+        if delta_count == 0:
+            delta_avg = 0
+        else:
+            delta_avg = float(delta_sum) / float(delta_count)
+
+        # publish raw data
+        if self.config['long_running_detail']:
+            self.publish_gauge(total_sum_name, total_sum)
+            self.publish_gauge(total_count_name, total_count)
+
+        # publish averages
+        self.publish_gauge(delta_avg_name, delta_avg, 6)
 
-        return json_data
+    def _ceph_time_to_seconds(self, val):
+        """Convert Ceph time format into seconds.  Older Ceph
+           versions output times as a string, while newer
+           versions output a float (which we pass through)
 
-    def _publish_stats(self, counter_prefix, stats):
-        """Given a stats dictionary from _get_stats_from_socket,
-        publish the individual values.
+        :param val: string in format "seconds.nanoseconds" or
+                    floating point number.
+
+        Returns:
+            Time in seconds as a floating point number.
         """
+        if isinstance(val, basestring):
+            sec, nsec = map(lambda v: long(v), val.split("."))
+            return float(sec * _NSEC_PER_SEC + nsec) / float(_NSEC_PER_SEC)
+        else:
+            return val
+
+    def _get_byte_metrics(self, name, metric_value):
+        """Return list of metrics derived from byte units.
+
+        Args:
+            name: the name of the metric
+            metric_value: the value of the metric in bytes
+
+        Returns:
+            List of (name, value) pairs for each unit.
+        """
+        assert name.endswith("bytes")
+        result = []
+        for unit in self.config['byte_unit']:
+            new_value = diamond.convertor.binary.convert(
+                value=metric_value, oldUnit='byte', newUnit=unit)
+            new_name = name.__class__(name.replace("bytes", unit))
+            result.append((new_name, new_value))
+        return result
+
+    def _publish_stats(self, prefix, stats, schema, name_class):
+        """Publish a set of Ceph performance counters, including schema.
+
+        :param prefix: string prefixed to metric names
+        :param stats: dictionary containing performance counters
+        :param schema: performance counter schema
+        """
+        for path, stat_type in flatten_dictionary(schema):
+            # remove 'stat_type' component to get metric name
+            if path[-1] != 'type':
+                continue
```

Comment:
The perf counter schema http://docs.ceph.com/docs/master/dev/perf_counters/#schema actually contains elements beyond type e.g. 
{u'description': u'a version',
 u'nick': u'',
 u'type': 2}
seems like I need to get the ceph docs updated to reflect reality

We skip the other elements for two reasons 
1. it would result in us publishing individual stats more than once for each collection 
see: https://github.com/GregMeno/Diamond-1/blob/calamari/src/collectors/ceph/test/testceph.py#L225

2. we need the type value from the schema to determine how to publish the stat